- Agenda :
1. Feature Selection using Forward Selection method and backwards elimination method
2. Dimensionality Reduction (Reducing the number of features present in the data using techniques like PCA and LDA)


## Notes:
- Feature Selection and Dimensionality Reduction both of these concepts are related to:  
    a. reducing the total number of columns from the data and   
    b. improving the accuracy of the model

In [1]:
!pip install mlxtend

Error processing line 3 of C:\Users\AMANT\anaconda3\lib\site-packages\googleapis_common_protos-1.56.4-py3.10-nspkg.pth:

  Traceback (most recent call last):
    File "C:\Users\AMANT\anaconda3\lib\site.py", line 169, in addpackage
      exec(line)
    File "<string>", line 1, in <module>
    File "<frozen importlib._bootstrap>", line 562, in module_from_spec
  AttributeError: 'NoneType' object has no attribute 'loader'

Remainder of file ignored


## =============== Feature Selection ================

## Load the all the libraries

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

## Load the data

In [3]:
from sklearn.datasets import load_breast_cancer
data = load_breast_cancer()
data

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
         1.189e-01],
        [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
         8.902e-02],
        [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
         8.758e-02],
        ...,
        [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
         7.820e-02],
        [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
         1.240e-01],
        [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
         7.039e-02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
        0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
        1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
        1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
        1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0

In [4]:
df = pd.DataFrame(np.c_[data['data'], data['target']], columns = np.append(data['feature_names'], ['target']))
df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  target  
0          0.4601                  0.11890     0.0  
1          0.2750                  0.08902     0.0  
2          0.3613                  0.08758     0.0  
3          0.6638                  0.17300     0.0  
4          0.2364                  0.07678     0.0  

[5 rows x 31 columns]

In [5]:
df.shape

(569, 31)

## Classify whether the patient is suffering from malignant or begnin tumor : Classification problem

## Seperate X and y

In [6]:
X = df.drop('target', axis = 1)
y = df['target']

## Split the data into train test split

In [7]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

## Apply Logistic Regression

In [8]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr

LogisticRegression()

In [9]:
lr.fit(X_train, y_train)

LogisticRegression()

In [10]:
y_pred = lr.predict(X_test)
y_pred

array([0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 0., 1., 0., 1.,
       0., 0., 0., 0., 0., 1., 1., 0., 1., 1., 0., 1., 0., 1., 0., 1., 0.,
       1., 0., 1., 0., 1., 0., 0., 1., 0., 1., 0., 0., 1., 1., 1., 0., 0.,
       1., 0., 1., 1., 1., 1., 1., 1., 0., 0., 0., 1., 1., 0., 1., 0., 0.,
       0., 1., 0., 0., 1., 1., 0., 1., 1., 1., 1., 1., 0., 0., 0., 1., 0.,
       1., 1., 1., 0., 0., 1., 0., 0., 0., 1., 1., 0., 1., 1., 1., 1., 1.,
       1., 1., 0., 1., 0., 1., 0., 0., 1., 0., 0., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 0., 1., 0., 1., 0., 1., 1., 1., 0., 1., 1., 1., 1., 1.,
       1., 0., 0., 1., 1., 1., 0.])

In [11]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred)

0.9440559440559441

## Conclude : With 31 columns in the data the accuracy achieved is 94.40%

## Forward Feature Selection Method: 
- The idea is to get either same accuracy or slightly lesser accuracy or better accuracy on a condition that the number of features should be less

- For eg: Let's selection only 10 features out of 31 features and then check accuracy


#### working of FFS method
-- Its considers 0 features are important. 
-- It considers 1 feaures as important
-- It considers 2 features as important
.
.
.
Contd. till it gets the 10 most important features from the data

In [12]:
## Applying LogReg of only 10 features of the data
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr

LogisticRegression()

In [13]:
from mlxtend.feature_selection import SequentialFeatureSelector as SFS
sfs = SFS(lr, k_features = 10, forward = True, scoring = 'accuracy')
sfs

SequentialFeatureSelector(estimator=LogisticRegression(), k_features=10,
                          scoring='accuracy')

In [14]:
sfs1 = sfs.fit(X_train, y_train)
sfs1

SequentialFeatureSelector(estimator=LogisticRegression(), k_features=(10, 10),
                          scoring='accuracy')

## Getting the names of those 10 most important features that were selected

In [15]:
col_index = sfs1.k_feature_idx_
col_index = list(col_index)
col_index

[0, 1, 2, 4, 5, 6, 18, 21, 22, 28]

In [16]:
X_train = pd.DataFrame(X_train, columns = df.drop('target', axis = 1).columns)
X_train.columns[col_index]

Index(['mean radius', 'mean texture', 'mean perimeter', 'mean smoothness',
       'mean compactness', 'mean concavity', 'symmetry error', 'worst texture',
       'worst perimeter', 'worst symmetry'],
      dtype='object')

In [21]:
X_train

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
293       11.850         17.46           75.54      432.7          0.08372   
332       11.220         19.86           71.94      387.3          0.10540   
565       20.130         28.25          131.20     1261.0          0.09780   
278       13.590         17.84           86.24      572.3          0.07948   
489       16.690         20.20          107.10      857.6          0.07497   
..           ...           ...             ...        ...              ...   
277       18.810         19.98          120.90     1102.0          0.08923   
9         12.460         24.04           83.97      475.9          0.11860   
359        9.436         18.32           59.82      278.6          0.10090   
192        9.720         18.22           60.73      288.1          0.06950   
559       11.510         23.93           74.52      403.5          0.09261   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
293           0.05642        0.026880             0.022800         0.1875   
332           0.06779        0.005006             0.007583         0.1940   
565           0.10340        0.144000             0.097910         0.1752   
278           0.04052        0.019970             0.012380         0.1573   
489           0.07112        0.036490             0.023070         0.1846   
..                ...             ...                  ...            ...   
277           0.05884        0.080200             0.058430         0.1550   
9             0.23960        0.227300             0.085430         0.2030   
359           0.05956        0.027100             0.014060         0.1506   
192           0.02344        0.000000             0.000000         0.1653   
559           0.10210        0.111200             0.041050         0.1388   

     mean fractal dimension  ...  worst radius  worst texture  \
293                 0.05715  ...        13.060          25.75   
332                 0.06028  ...        11.980          25.78   
565                 0.05533  ...        23.690          38.25   
278                 0.05520  ...        15.500          26.10   
489                 0.05325  ...        19.180          26.56   
..                      ...  ...           ...            ...   
277                 0.04996  ...        19.960          24.30   
9                   0.08243  ...        15.090          40.68   
359                 0.06959  ...        12.020          25.02   
192                 0.06447  ...         9.968          20.83   
559                 0.06570  ...        12.480          37.16   

     worst perimeter  worst area  worst smoothness  worst compactness  \
293            84.35       517.8           0.13690            0.17580   
332            76.91       436.1           0.14240            0.09669   
565           155.00      1731.0           0.11660            0.19220   
278            98.91       739.1           0.10500            0.07622   
489           127.30      1084.0           0.10090            0.29200   
..               ...         ...               ...                ...   
277           129.00      1236.0           0.12430            0.11600   
9              97.65       711.4           0.18530            1.05800   
359            75.79       439.6           0.13330            0.10490   
192            62.25       303.8           0.07117            0.02729   
559            82.28       474.2           0.12980            0.25170   

     worst concavity  worst concave points  worst symmetry  \
293          0.13160               0.09140          0.3101   
332          0.01335               0.02022          0.3292   
565          0.32150               0.16280          0.2572   
278          0.10600               0.05185          0.2335   
489          0.24770               0.08737          0.4677   
..               ...                   ...             ...   
277          0.22100               0.12940         

In [17]:
X_train = X_train.iloc[:, col_index]
X_train

mean radius  mean texture  mean perimeter  mean smoothness  \
293       11.850         17.46           75.54          0.08372   
332       11.220         19.86           71.94          0.10540   
565       20.130         28.25          131.20          0.09780   
278       13.590         17.84           86.24          0.07948   
489       16.690         20.20          107.10          0.07497   
..           ...           ...             ...              ...   
277       18.810         19.98          120.90          0.08923   
9         12.460         24.04           83.97          0.11860   
359        9.436         18.32           59.82          0.10090   
192        9.720         18.22           60.73          0.06950   
559       11.510         23.93           74.52          0.09261   

     mean compactness  mean concavity  symmetry error  worst texture  \
293           0.05642        0.026880         0.01792          25.75   
332           0.06779        0.005006         0.04243          25.78   
565           0.10340        0.144000         0.01898          38.25   
278           0.04052        0.019970         0.01344          26.10   
489           0.07112        0.036490         0.01961          26.56   
..                ...             ...             ...            ...   
277           0.05884        0.080200         0.01930          24.30   
9             0.23960        0.227300         0.01789          40.68   
359           0.05956        0.027100         0.01942          25.02   
192           0.02344        0.000000         0.03799          20.83   
559           0.10210        0.111200         0.01488          37.16   

     worst perimeter  worst symmetry  
293            84.35          0.3101  
332            76.91          0.3292  
565           155.00          0.2572  
278            98.91          0.2335  
489           127.30          0.4677  
..               ...             ...  
277           129.00          0.2567  
9              97.65          0.4366  
359            75.79          0.2454  
192            62.25          0.1909  
559            82.28          0.2112  

[426 rows x 10 columns]

In [18]:
lr.fit(X_train, y_train)

LogisticRegression()

In [20]:
X_test = X_test.iloc[:, col_index]
X_test

mean radius  mean texture  mean perimeter  mean smoothness  \
512        13.40         20.52           88.64          0.11060   
457        13.21         25.25           84.10          0.08791   
439        14.02         15.66           89.59          0.07966   
298        14.26         18.17           91.22          0.06576   
37         13.03         18.42           82.61          0.08983   
..           ...           ...             ...              ...   
236        23.21         26.97          153.50          0.09509   
113        10.51         20.19           68.64          0.11220   
527        12.34         12.27           78.94          0.09003   
76         13.53         10.94           87.91          0.12910   
162        19.59         18.15          130.70          0.11200   

     mean compactness  mean concavity  symmetry error  worst texture  \
512           0.14690         0.14450         0.01705          29.66   
457           0.05205         0.02772         0.01347          34.23   
439           0.05581         0.02087         0.01171          19.31   
298           0.05220         0.02475         0.01103          25.26   
37            0.03766         0.02562         0.02671          22.81   
..                ...             ...             ...            ...   
236           0.16820         0.19500         0.01590          34.51   
113           0.13030         0.06476         0.02287          22.75   
527           0.06307         0.02958         0.01129          19.27   
76            0.10470         0.06877         0.02572          12.49   
162           0.16660         0.25080         0.01686          26.39   

     worst perimeter  worst symmetry  
512           113.30          0.3585  
457            91.29          0.2444  
439            96.53          0.2136  
298           105.80          0.2636  
37             84.46          0.1987  
..               ...             ...  
236           206.00          0.3103  
113            72.62          0.2383  
527            87.22          0.3110  
76             91.36          0.2710  
162           174.90          0.3643  

[143 rows x 10 columns]

In [21]:
y_pred = lr.predict(X_test)

In [22]:
y_pred

array([0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 0., 1., 0., 1.,
       0., 0., 0., 0., 0., 1., 1., 0., 1., 1., 0., 1., 0., 1., 0., 1., 0.,
       1., 0., 1., 0., 1., 0., 0., 1., 0., 1., 0., 0., 1., 1., 1., 0., 0.,
       0., 0., 1., 1., 1., 1., 1., 1., 0., 0., 0., 1., 1., 0., 1., 0., 0.,
       0., 1., 0., 0., 1., 1., 0., 1., 1., 1., 1., 1., 0., 0., 0., 1., 0.,
       1., 1., 1., 0., 0., 1., 0., 0., 0., 1., 1., 0., 1., 1., 1., 1., 1.,
       1., 1., 0., 1., 0., 1., 0., 0., 1., 0., 0., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 0., 1., 0., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1., 1.,
       0., 0., 0., 1., 1., 1., 0.])

In [23]:
accuracy_score(y_pred, y_test)

0.951048951048951

## Conclude : With only 10 features the accuracy achieved is 95.10%

# Method 2 : Backward Elimination Method

Working of Backward Elimination Method:
- It uses ML algorithm to select the most important features from the data
- It starts with all features included in the model and then eliminates the least relevant features one by one

In [25]:
from mlxtend.feature_selection import SequentialFeatureSelector as SFS
sfs = SFS(lr, k_features = 12, forward = False, scoring = 'accuracy')
sfs

SequentialFeatureSelector(estimator=LogisticRegression(), forward=False,
                          k_features=12, scoring='accuracy')

In [26]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

In [29]:
sfs_back = sfs.fit(X_train, y_train)
sfs_back

SequentialFeatureSelector(estimator=LogisticRegression(), forward=False,
                          k_features=(12, 12), scoring='accuracy')

In [30]:
## Getting the names of all the 12 important features selected by Backward elimination technique

In [31]:
sfs_back.k_feature_idx_

(0, 1, 4, 5, 6, 7, 8, 12, 20, 22, 23, 25)

In [33]:
## Getting the names

X_train = pd.DataFrame(X_train, columns = df.drop('target', axis = 1).columns)
X_train.columns[list(sfs_back.k_feature_idx_)]

Index(['mean radius', 'mean texture', 'mean smoothness', 'mean compactness',
       'mean concavity', 'mean concave points', 'mean symmetry',
       'perimeter error', 'worst radius', 'worst perimeter', 'worst area',
       'worst compactness'],
      dtype='object')

In [34]:
X_train = X_train.iloc[:, list(sfs_back.k_feature_idx_)]
X_train

mean radius  mean texture  mean smoothness  mean compactness  \
293       11.850         17.46          0.08372           0.05642   
332       11.220         19.86          0.10540           0.06779   
565       20.130         28.25          0.09780           0.10340   
278       13.590         17.84          0.07948           0.04052   
489       16.690         20.20          0.07497           0.07112   
..           ...           ...              ...               ...   
277       18.810         19.98          0.08923           0.05884   
9         12.460         24.04          0.11860           0.23960   
359        9.436         18.32          0.10090           0.05956   
192        9.720         18.22          0.06950           0.02344   
559       11.510         23.93          0.09261           0.10210   

     mean concavity  mean concave points  mean symmetry  perimeter error  \
293        0.026880             0.022800         0.1875            1.234   
332        0.005006             0.007583         0.1940            1.959   
565        0.144000             0.097910         0.1752            5.203   
278        0.019970             0.012380         0.1573            1.683   
489        0.036490             0.023070         0.1846            1.775   
..              ...                  ...            ...              ...   
277        0.080200             0.058430         0.1550            2.363   
9          0.227300             0.085430         0.2030            2.039   
359        0.027100             0.014060         0.1506            3.267   
192        0.000000             0.000000         0.1653            2.230   
559        0.111200             0.041050         0.1388            1.936   

     worst radius  worst perimeter  worst area  worst compactness  
293        13.060            84.35       517.8            0.17580  
332        11.980            76.91       436.1            0.09669  
565        23.690           155.00      1731.0            0.19220  
278        15.500            98.91       739.1            0.07622  
489        19.180           127.30      1084.0            0.29200  
..            ...              ...         ...                ...  
277        19.960           129.00      1236.0            0.11600  
9          15.090            97.65       711.4            1.05800  
359        12.020            75.79       439.6            0.10490  
192         9.968            62.25       303.8            0.02729  
559        12.480            82.28       474.2            0.25170  

[426 rows x 12 columns]

In [35]:
X_test = X_test.iloc[:, list(sfs_back.k_feature_idx_)]
X_test

mean radius  mean texture  mean smoothness  mean compactness  \
512        13.40         20.52          0.11060           0.14690   
457        13.21         25.25          0.08791           0.05205   
439        14.02         15.66          0.07966           0.05581   
298        14.26         18.17          0.06576           0.05220   
37         13.03         18.42          0.08983           0.03766   
..           ...           ...              ...               ...   
236        23.21         26.97          0.09509           0.16820   
113        10.51         20.19          0.11220           0.13030   
527        12.34         12.27          0.09003           0.06307   
76         13.53         10.94          0.12910           0.10470   
162        19.59         18.15          0.11200           0.16660   

     mean concavity  mean concave points  mean symmetry  perimeter error  \
512         0.14450              0.08172         0.2116           3.0930   
457         0.02772              0.02068         0.1619           1.3140   
439         0.02087              0.02652         0.1589           1.6060   
298         0.02475              0.01374         0.1635           1.6610   
37          0.02562              0.02923         0.1467           1.1700   
..              ...                  ...            ...              ...   
236         0.19500              0.12370         0.1909           7.2470   
113         0.06476              0.03068         0.1922           2.0410   
527         0.02958              0.02647         0.1689           0.7714   
76          0.06877              0.06556         0.2403           2.6520   
162         0.25080              0.12860         0.2027           4.7920   

     worst radius  worst perimeter  worst area  worst compactness  
512         16.41           113.30       844.4            0.38560  
457         14.35            91.29       632.9            0.10630  
439         14.91            96.53       688.9            0.10170  
298         16.22           105.80       819.7            0.21670  
37          13.30            84.46       545.9            0.04619  
..            ...              ...         ...                ...  
236         31.01           206.00      2944.0            0.41260  
113         11.16            72.62       374.4            0.20490  
527         13.61            87.22       564.9            0.20740  
76          14.08            91.36       605.5            0.13790  
162         26.73           174.90      2232.0            0.38460  

[143 rows x 12 columns]

In [36]:
lr.fit(X_train, y_train)

LogisticRegression()

In [37]:
y_pred = lr.predict(X_test)

In [38]:
accuracy_score(y_pred, y_test)

0.951048951048951

## Conclude : Using Backward Elimination and selecting only 12 most important columns we get an accuracy of 95.10%

## Exhaustive Search Feature Selection:

- Combo of Forward Selection and Backward Elimination
- Where it can helpful?
- On Huge dataset with 10000 rows and 175 columns, When you use Forward Selection will take time, when you use Backward Elimination it will take more time and processing compared to Forward Selection, to overcome these disadvantages you can go with Exhaustive Search Method

In [39]:
from mlxtend.feature_selection import ExhaustiveFeatureSelector as EFS
efs = EFS(lr, min_features = 1, max_features = 8 ,scoring = 'accuracy')
efs

ExhaustiveFeatureSelector(estimator=LogisticRegression(), max_features=8)

In [40]:
FS : You specify 10 imp features in SFS algorithm - 10 features must be selected even though the data does 
    not has 10 imp features
BE :You specify 12 imp features in SFS algorithm - 12 features must be selection even though the data does 
    not has 12 imp features
ES : You specify the range 1 - 10 features in EFS algorithm - any no of features can be selected in the range 
    specified. Not necessary to select all the 10 features

SyntaxError: invalid syntax (3511306209.py, line 1)

In [41]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

In [ ]:
efs.fit(X_train, y_train)
efs